enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,407 @@
+
+```yaml
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
+metadata:
+  name: openshift-node-workload-partitioning-custom
+spec:
+  cpu:
+    isolated: 2-3
+    reserved: 0,1
+    # New addition
+    workloads:
+      enablePinning: true
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Changes to NTO
+
+The NTO PerformanceProfile will be updated to support a new flag which will
+toggle the workload pinning to the `isolated` cores. The idea here being to
+simplify the approach for how customers set this configuration. With PAO being
+part of NTO now ([see here for more info](../node-tuning/pao-in-nto.md)) this
+affords us the chance to consolidate the configuration for `kubelet` and `crio`.
+
+We will modify the code path that generates the [new machine
+config](https://github.com/openshift/cluster-node-tuning-operator/blob/a780dfe07962ad07e4d50c852047ef8cf7b287da/pkg/performanceprofile/controller/performanceprofile/components/machineconfig/machineconfig.go#L91-L127)
+using the performance profile. With the new `workloads.enablePinning` flag we
+will add the configuration for `crio` and `kubelet` to the final machine config
+manifest. Then the existing code path will apply the change as normal.
+
+#### API Server Admission Hook
+
+The existing admission hook has 4 checks when it comes to workload pinning.
+
+1. Check if `pod` is a static pod
+   - Skips modification attempt if it is static.
+1. Checks if currently running cluster topology is Single Node
+   - Skips modification if it is anything other than Single Node
+1. Checks if all running nodes are managed
+   - Skips modification if any of the nodes are not managed
+1. Checks what resource limits and requests are set on the pod
+   - Skips modification if QoS is guaranteed or both limits and requests are set
+   - Skips modification if after update the QoS is changed
+
+We will need to alter the code in the admission controller to remove the check
+for Single Node Topology, the other configurations should remain untouched.
+
+### Risks and Mitigations
+
+The sames risks and mitigations highlighted in [Management Workload
+Partitioning](management-workload-partitioning.md) apply to this enhancement as
+well.
+
+We need to make it very clear to customers that this feature is supported as a
+day 0 configuration and day n+1 alterations are not be supported with this
+enhancement. Part of that messaging should involve a clear indication that this
+should be a cluster wide feature.
+
+A risk we can run into is that a customer can apply a CPU set that is too small
+or out of bounds can cause problems such as extremely poor performance or start
+up errors. Mitigation of this scenario will be to provide proper guidance and
+guidelines for customers who enable this enhancement. Furthermore, the
+performance team would need to be reached out to for more specific information
+around upper and lower bounds of CPU sets for running an Openshift cluster.
```

Comment:
We can add a check for the lower bounds, and as it stands the PerformanceProfile API does a check for empty values in the CPUSets. However, I'm not sure if that is a desired validation, since the typical customer for this type of feature is wanting a smaller footprint for the management workloads. If we do a validation on CPUSet size, I think we'll need to maintain some hard coded value or allow some configurable value for what is deemed a good size for a CPUSet.

On the upper bounds I do not know what is available for validation, since it would seem that the Performance Profile Controller would need to be aware of the nodes in the machine config pool to validate upper bounds size for the CPUSets.

@mrunalp and @haircommander Keep me honest, what are your thoughts on this?

Reply:
Sounds good, updated this paragraph to make mention of it, as well as recorded our conclusion in a goal and the variation.